File: solutions/java/src/main/java/com/epi/IndirectSort.java
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.
package com.epi;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;
import java.util.Random;
import java.util.Scanner;

public class IndirectSort {
  // @include
  public static void indirectSort(String fileName) throws IOException {
    // Stores file records into A.
    Scanner ifs = null;
    List<Integer> A = new ArrayList<>();
    try {
      ifs = new Scanner(new File(fileName));
      while (ifs.hasNextInt()) {
        A.add(ifs.nextInt());
      }
    } finally {
      if (ifs != null) {
        ifs.close();
      }
    }

    // Indirectly sorts file.
    Collections.sort(A);

    // Outputs file.
    PrintWriter ofs = null;
    try {
      ofs = new PrintWriter(new FileWriter(fileName));
      for (Integer a : A) {
        ofs.println(a);
      }
    } finally {
      if (ofs != null) {
        ofs.close();
      }
    }
  }
  // @exclude

  public static void main(String[] args) throws IOException {
    Random rnd = new Random();
    for (int times = 0; times < 1000; ++times) {
      System.out.println("times = " + times);
      int n = 0;
      if (args.length == 1) {
        n = Integer.parseInt(args[0]);
      } else {
        n = rnd.nextInt(10000) + 1;
      }
      List<Integer> A = new ArrayList<>();
      for (int i = 0; i < n; i++) {
        A.add(rnd.nextInt(999999 + 1));
      }

      PrintWriter ofs = null;
      try {
        ofs = new PrintWriter(new FileWriter("input.txt"));
        for (Integer a : A) {
          ofs.println(a);
        }
      } finally {
        if (ofs != null) {
          ofs.close();
        }
      }
      indirectSort("input.txt");

      Scanner ifs = null;
      File input = new File("input.txt");
      try {
        ifs = new Scanner(input);
        A.clear();
        while (ifs.hasNextInt()) {
          A.add(ifs.nextInt());
        }
      } finally {
        if (ifs != null) {
          ifs.close();
        }
      }

      List<Integer> Adup = new ArrayList<>(A);
      Collections.sort(Adup);
      assert(Adup.equals(A));
      input.delete();
    }
  }
}

File: solutions/cpp/indirect_sort.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <fstream>
#include <iostream>
#include <iterator>
#include <random>
#include <string>
#include <vector>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::ifstream;
using std::ofstream;
using std::random_device;
using std::string;
using std::uniform_int_distribution;
using std::vector;

// @include
void IndirectSort(const string& file_name) {
  // Stores file records into A.
  ifstream ifs(file_name.c_str());
  vector<int> A;
  int x;
  while (ifs >> x) {
    A.emplace_back(x);
  }

  // Initializes P.
  vector<const int*> P;
  for (int& a : A) {
    P.emplace_back(&a);
  }

  // Indirectly sorts file.
  sort(begin(P), end(P), [](const int* a, const int* b) { return *a < *b; });

  // Outputs file.
  ofstream ofs(file_name.c_str());
  for (const int* p : P) {
    ofs << *p << endl;
  }
}
// @exclude

int main(int argc, char* argv[]) {
  default_random_engine gen((random_device())());
  for (int times = 0; times < 1000; ++times) {
    cout << "times = " << times << endl;
    int n;
    if (argc == 2) {
      n = atoi(argv[1]);
    } else {
      uniform_int_distribution<int> dis(1, 10000);
      n = dis(gen);
    }
    vector<int> A;
    generate_n(back_inserter(A), n, [&] {
      uniform_int_distribution<int> dis(0, 999999);
      return dis(gen);
    });
    ofstream ofs("input.txt");
    for (int a : A) {
      ofs << a << endl;
    }

    IndirectSort("input.txt");
    ifstream ifs("input.txt");
    A.clear();
    int x;
    while (ifs >> x) {
      A.emplace_back(x);
    }
    assert(is_sorted(A.begin(), A.end()));
    remove("input.txt");
  }
  return 0;
}
